Guard returns of limited-availability cases in init(rawValue:)

beccadax opened the pull request:
Suppose you tried to compile something like this with a minimum deployment target of iOS 9:

```swift
enum E: Int {
  @available(iOS 10, tvOS 10, *) case c
}
```

The compiler would try to synthesize an `init(rawValue:)` like this one:

```swift
  init?(rawValue: Int) {
    switch rawValue {
    case 0:
      return E.c
    default:
      return nil
    }
  }
```

However, that code would fail to compile; the availability checker would correctly notice that `E.c` is not available on all OSes where this code can run and this use was not guarded with an `#available` check, so it emits an error.

This change modifies `RawRepresentable` auto-derivation to synthesize this instead:

```swift
    case 0:
      guard #available(iOS 10, *) else { return nil }
      return E.c
```

That keeps `init(rawValue:)` from returning cases that aren't supported on the platform it's running on.

However, even though the generated code is now correct, the compiler still diagnoses an error. The availability checker is `SourceLoc`-based and skips over code with invalid `SourceLoc`s; changing that would require either a significant redesign of the availability checker (possibly with something like ASTScope) or augmenting `SourceLoc`s to represent the locations of synthesized code (something I explored in my free time in #22162—it's interesting, but definitely not ready). To get this fix out the door, I've added a targeted hack which disables potential availability checking in implicit `init(rawValue:)` implementations only.

Fixes rdar://problem/47488617.

beccadax committed:
[NFC] Refactor diagnoseDeclAvailability() flags

The function took three random boolean parameters; there’s about to be a fourth. This wasn’t going to scale.

beccadax committed:
Add #available guards to init(rawValue:)

When a case has an @available(currentPlatform, introduced: x.y) attribute, the auto-derived init(rawValue:) will now include a “guard #available(currentPlatform x.y, *) else { return nil }” check to prevent the new case from being returned.

This requires a targeted hack to disable @available(introduced:) checks in auto-derived init(rawValue:) implementations; the availability checker is SourceLoc-based, so it can’t tell that the synthesized `guard` covers the use of the case. I’m not happy about that, but fixing the deeper problem is a much larger task than I can take on in one day.

beccadax committed:
Don’t run #available synthesis test on Linux

brentdax commented:
@swift-ci please test

devincoughlin commented on lib/Sema/DerivedConformanceRawRepresentable.cpp:
```diff
@@ -245,23 +250,78 @@ deriveBodyRawRepresentable_init(AbstractFunctionDecl *initDecl, void *) {
+    // If there's an @available attribute specifying when the case was
+    // introduced, generate an early return on unavailability, e.g.
+    // "guard #available(iOS 9, *) else { return nil }"
+    for (auto *attr : elt->getAttrs().getAttributes<AvailableAttr>()) {
```
I suggest moving the logic that iterates over the active attributes to Attr.cpp as `DeclAttributes::getIntroduced()` to be analogous to `Decl::Attributes::getDeprecated()`. This will put all the attribute traversal logic in one place.

brentdax replied:
`getUnavailable()` skips over `AvailableVersionComparison::PotentiallyUnavailable` attributes; is that a bug in the function or a necessary difference?

devincoughlin replied:
I think I told you the wrong thing in my second comment above. I do think you should probably move your code to Attr.cpp as DeclAttributes::getIntroduced() rather than change getUnavailable().

devincoughlin reviewed (approved):
It is really great to see the refactoring here.

Looks good to me with the comments inline addressed

jrose-apple commented on test/SILGen/enum_raw_representable_available.swift:
```diff
@@ -0,0 +1,21 @@
+
+
+// CHECK-RESILIENT-DAG: sil [ossa] @$s32enum_raw_representable_available1EO0B5ValueACSgSi_tcfC
+// CHECK-RESILIENT-DAG: sil [ossa] @$s32enum_raw_representable_available1EO0B5ValueSivg
```
SIL is the right way to test the behavior of generated AST code (if you don't want to go all the way to a regression test).

jrose-apple replied:
…but now that I've read the code, yes, we should be testing that the generated SIL is what we expect.

Rather than having separate CHECK and CHECK-RESILIENT sections, I suggest just optionally matching the "serialized" attribute, so that we can check that the body is the same either way.

jrose-apple commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -2301,6 +2301,19 @@ class AvailabilityWalker : public ASTWalker {
+           init->isImplicit() &&
+           init->getParameters()->size() == 1 &&
+           init->getParameters()->get(0)->getArgumentName() ==
+                   DC->getASTContext().Id_rawValue;
```
Nitpick: `TC` provides faster access to the ASTContext than an AST node.

jrose-apple commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -2320,8 +2333,20 @@ class AvailabilityWalker : public ASTWalker {
+        // on older versions, but the availability checker can't verify that
+        // because the synthesized code uses invalid SourceLocs. Don't diagnose
+        // these errors; instead, take it on faith that
+        // DerivedConformanceRawRepresentable will do the right thing.
```
\*sigh\* I don't love it because if we introduce *another* availability mistake we won't see it, but it's probably the best we can do right now.

jrose-apple commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -2504,46 +2527,45 @@ class AvailabilityWalker : public ASTWalker {
-                                bool ForInout) const {
+                                DeclAvailabilityFlags Flags) const {
+    Flags &= DeclAvailabilityFlag::ForInout;
+    Flags |= DeclAvailabilityFlag::ContinueOnPotentialUnavailability;
```
This changes the behavior for AllowPotentiallyUnavailableProtocol. Do we need a test for that?

jrose-apple replied:
No, that's sufficient justification. :-)

jrose-apple commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -2582,20 +2604,25 @@ bool AvailabilityWalker::diagAvailability(const ValueDecl *D, SourceRange R,
+        && isa<ProtocolDecl>(D))
+    return false;
+
+  if (Flags.contains(DeclAvailabilityFlag::AllowPotentiallyUnavailable))
```
Nitpick: I think it makes sense to check these in the opposite order, to avoid doing work.

brentdax commented on lib/Sema/DerivedConformanceRawRepresentable.cpp:
```diff
@@ -245,23 +250,78 @@ deriveBodyRawRepresentable_init(AbstractFunctionDecl *initDecl, void *) {
+        // guardStmt = "guard \(vailableInfo) else \(earlyReturnBody)"
+        SmallVector<StmtConditionElement, 1>
+        conds{StmtConditionElement(availableInfo)};
+        auto guardStmt = new (C) GuardStmt(SourceLoc(), C.AllocateCopy(conds),
```
Oddly, I can't seem to make this work without explicitly spelling out `ArrayRef<StmtConditionElement>(...)`, but I can at least use a plain C array to reduce the overhead.

beccadax committed:
Further improve AvailabilityWalker changes

Access context more quickly; document each flag’s purpose; perform a faster test before a slower one.

beccadax committed:
Add accessor with potentially unavailable attrs

DeclAttributes::getUnavailable() only cares about attributes which make a declaration definitely unavailable, but you sometimes need a version which will also return a potentially unavailable (i.e. “introduced:”) attribute. This adds that.

beccadax committed:
Refactor availability-checking code to handle more cases

Moves a lot of it into helper functions and types so it doesn’t disrupt the main flow of the code so much. Also makes it handle always-unavailable and obsolete cases (by skipping them).

brentdax commented:
@swift-ci please test

brentdax commented:
While updating this PR to address the review comments, I decided we really ought to skip over unavailable and obsolete cases in `init(rawValue:)`. These currently generate weird errors at `<unknown>:0` too, so skipping them should be an improvement.

brentdax commented:
@swift-ci please test compiler performance

brentdax commented:
(Running a compiler performance test to look at issues with those tests, not really because I think this PR will affect performance.)

beccadax committed:
Lower versions in new enum availability test

32-bit iOS Simulator tests are compiled with an iOS 7 depolyment target, so the “always available” case wasn’t available there, causing a spurious test failure.

brentdax commented:
@devincoughlin @jrose-apple Do you mind looking this over again? The change to AST/Attr.{h,cpp} is new, the change to Sema/DerivedConformanceRawRepresentable.cpp has been refactored and implements different behavior, and the test interrogates the SIL much more aggressively.

brentdax commented:
@swift-ci please smoke test compiler performance

brentdax commented:
(Another performance test to test the run_cperf script.)

brentdax commented:
@swift-ci please test compiler performance

jrose-apple commented on lib/AST/Attr.cpp:
```diff
@@ -131,6 +131,48 @@ DeclAttributes::isUnavailableInSwiftVersion(
 }
 
+const AvailableAttr *DeclAttributes::getPotentiallyUnavailable(
+                          const ASTContext &ctx) const {
```
Nitpick: LLVM style says to just double-indent this, rather than trying to line it up with something. (Or whatever clang-format does on our repo.)

brentdax replied:
This particular decl is probably best wrapped after the return type, but thank you—I've been struggling to figure out what to do when a function name gets this long.

jrose-apple commented on lib/AST/Attr.cpp:
```diff
@@ -131,6 +131,48 @@ DeclAttributes::isUnavailableInSwiftVersion(
+        return AvAttr;
+
+      switch (AvAttr->getVersionAvailability(ctx)) {
+        case AvailableVersionComparison::Available:
```
Nitpick: cases shouldn't be indented past the `switch` in LLVM style.

jrose-apple commented on lib/Sema/DerivedConformanceRawRepresentable.cpp:
```diff
@@ -188,17 +188,98 @@ static VarDecl *deriveRawRepresentable_raw(DerivedConformance &derived) {
+
+    // availableInfo = "#available(\(platformSpec), \(otherSpec))"
+    auto availableInfo = PoundAvailableInfo::create(C, SourceLoc(),
+                                              { platformSpec, otherSpec },
```
Same kind of indentation comments: if something doesn't fit, just wrap it to be double-indented rather than pushing it over from the right.

jrose-apple commented on lib/Sema/DerivedConformanceRawRepresentable.cpp:
```diff
@@ -188,17 +188,98 @@ static VarDecl *deriveRawRepresentable_raw(DerivedConformance &derived) {
+/// be available, returns true. If it will sometimes be available, adds
+/// information about the runtime check needed to ensure it is available to
+/// \c versionCheck and returns true.
+static bool checkAvailability(EnumElementDecl* elt, ASTContext &C,
```
Nitpick: `elt` could be `const`.

jrose-apple commented on lib/Sema/DerivedConformanceRawRepresentable.cpp:
```diff
@@ -188,17 +188,98 @@ static VarDecl *deriveRawRepresentable_raw(DerivedConformance &derived) {
+  if (availability != AvailableVersionComparison::PotentiallyUnavailable)
+    return false;
+
+  // It's conditionally available; create a version constraint and return true.
```
Can you assert that it's platform availability in this case, and not some kind of untestable weak linking we invent in the future?

jrose-apple commented on test/SILGen/enum_raw_representable_available.swift:
```diff
@@ -0,0 +1,67 @@
+// RUN: %target-swift-emit-silgen -emit-sorted-sil -enable-resilience %s | %FileCheck %s
+
+// Really just requires a platform with meaningful runtime #available() tests.
+// REQUIRES: objc_interop
```
I think it would be fine to limit this to "just macOS" or something, rather than get bitten when Apple introduces a new platform some day. `REQUIRES: OS=macosx`

jrose-apple commented on test/SILGen/enum_raw_representable_available.swift:
```diff
@@ -0,0 +1,67 @@
+
+// CHECK-LABEL: sil {{(\[serialized\] )?}}[ossa] @$s32enum_raw_representable_available1EO0B5ValueACSgSi_tcfC
+
+// CHECK-NOT: integer_literal $Builtin.IntLiteral, -4000
```
Rather than insert this everywhere, I suggest just doing two passes with FileCheck. (You can see examples of other tests doing this with `-check-prefix NEGATIVE`.)

jrose-apple reviewed (commented):
Looks good to me; all my comments are nitpicks. I think Devin should still take a look too, though.

beccadax committed:
[NFC] Style/assertion/test design improvements

beccadax committed:
Test init(rawValue:) behavior for obsolete cases

Also touches the other version numbers so I can use realistic, but far-future, ones.

devincoughlin commented:
LGTM. It's great to see more tests!

brentdax commented:
@swift-ci please test

brentdax commented:
Test failure is unrelated; fixing in #22531.

brentdax commented:
@swift-ci please test
